File: mvpLibrary/src/main/java/cn/cbsd/mvplibrary/mvp/XFragment.java
package cn.cbsd.mvplibrary.mvp;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentTransaction;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.fengchen.uistatus.UiStatusController;
import com.tbruyelle.rxpermissions2.RxPermissions;
import com.trello.rxlifecycle2.components.support.RxFragment;

import butterknife.Unbinder;
import cn.cbsd.mvplibrary.CommonConfig;
import cn.cbsd.mvplibrary.event.BusProvider;
import cn.cbsd.mvplibrary.kit.KnifeKit;

/**
 * Created by wanglei on 2016/12/29.
 */

public abstract class XFragment<P extends IPresent> extends RxFragment implements IView<P> {

    private VDelegate vDelegate;
    private P p;
    protected Activity context;
    private View rootView;
    protected LayoutInflater layoutInflater;
    private RxPermissions rxPermissions;
    private Unbinder unbinder;

    private static final String STATE_SAVE_IS_HIDDEN = "STATE_SAVE_IS_HIDDEN";
    private UiStatusController mUiController;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState != null) {
            boolean isSupportHidden = savedInstanceState.getBoolean(STATE_SAVE_IS_HIDDEN);

            FragmentTransaction ft = getFragmentManager().beginTransaction();
            if (isSupportHidden) {
                ft.hide(this);
            } else {
                ft.show(this);
            }
            ft.commit();
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        //保存隐藏的状态
        outState.putBoolean(STATE_SAVE_IS_HIDDEN, isHidden());
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        layoutInflater = inflater;
        if (rootView == null && getLayoutId() > 0) {
            rootView = inflater.inflate(getLayoutId(), null);
            bindUI(rootView);
        } else {
            ViewGroup viewGroup = (ViewGroup) rootView.getParent();
            if (viewGroup != null) {
                viewGroup.removeView(rootView);
            }
        }

        if (useDefaultUiState()) {
            mUiController = UiStatusController.get();
            rootView = mUiController.bindFragment(rootView);
        }
        return rootView;
    }

    /**
     * 获取默认的UiState
     * @return
     */
    public UiStatusController getDefaultUiController() {
        return mUiController;
    }

    /**
     * 使用默认的UiState，即添加在界面的顶层
     * @return
     */
    @Override
    public boolean useDefaultUiState() {
        return false;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (useEventBus()) {
            BusProvider.getBus().register(this);
        }
        bindEvent();
        initData(savedInstanceState);
    }

    @Override
    public void bindUI(View rootView) {
        unbinder = KnifeKit.bind(this, rootView);
    }

    public VDelegate getvDelegate() {
        if (vDelegate == null) {
            vDelegate = VDelegateBase.create(context);
        }
        return vDelegate;
    }

    protected P getP() {
        if (p == null) {
            p = newP();
            if (p != null) {
                p.attachV(this);
            }
        }
        return p;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof Activity) {
            this.context = (Activity) context;
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        context = null;
    }

    @Override
    public boolean useEventBus() {
        return false;
    }


    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (useEventBus()) {
            BusProvider.getBus().unregister(this);
        }
        if (getP() != null) {
            getP().detachV();
        }
        getvDelegate().destroy();

        p = null;
        vDelegate = null;
    }

    protected RxPermissions getRxPermissions() {
        rxPermissions = new RxPermissions(getActivity());
        rxPermissions.setLogging(CommonConfig.DEV);
        return rxPermissions;
    }

    @Override
    public int getOptionsMenuId() {
        return 0;
    }

    @Override
    public void bindEvent() {

    }

    @Override
    public void showLoading() {
        if (!getActivity().isFinishing())
            getvDelegate().showLoading("加载中...");
    }

    @Override
    public void showLoading(String msg) {
        if (getActivity()!=null && !getActivity().isFinishing()) {
            getvDelegate().showLoading(msg);
        }
    }

    @Override
    public void hideLoading() {
        getvDelegate().dismissLoading();
    }


}

File: mvpLibrary/src/main/java/cn/cbsd/mvplibrary/mvp/XFragment.kt
package cn.cbsd.mvplibrary.mvp

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import butterknife.Unbinder
import cn.cbsd.mvplibrary.CommonConfig
import cn.cbsd.mvplibrary.event.BusProvider
import cn.cbsd.mvplibrary.kit.KnifeKit
import com.fengchen.uistatus.UiStatusController
import com.tbruyelle.rxpermissions2.RxPermissions
import com.trello.rxlifecycle2.components.support.RxFragment
import kotlin.properties.Delegates

/**
 * Created by wanglei on 2016/12/29.
 */

abstract class XFragment : RxFragment(), IView {

    protected lateinit var context: Activity
    private val vDelegate: VDelegate by lazy { VDelegateBase(context) }
    private var rootView: View by Delegates.notNull()
    lateinit var myLayoutInflater: LayoutInflater

    val rxPermissions: RxPermissions by lazy {
        val rxPermissions = RxPermissions(context)
        rxPermissions.setLogging(CommonConfig.DEV)
        rxPermissions
    }
    private var unbinder: Unbinder? = null
    //private var p: P? = null

    /**
     * 获取默认的UiState
     * @return
     */
    lateinit var defaultUiController: UiStatusController

    override val optionsMenuId: Int
        get() = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (savedInstanceState != null) {
            val isSupportHidden = savedInstanceState.getBoolean(STATE_SAVE_IS_HIDDEN)

            val ft = fragmentManager!!.beginTransaction()
            if (isSupportHidden) {
                ft.hide(this)
            } else {
                ft.show(this)
            }
            ft.commit()
        }

    }

    override fun onSaveInstanceState(outState: Bundle) {
        //保存隐藏的状态
        outState.putBoolean(STATE_SAVE_IS_HIDDEN, isHidden)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        myLayoutInflater = inflater
        if (layoutId > 0) {
            rootView = inflater.inflate(layoutId, null)
            bindUI(rootView)
        } else {
            val viewGroup = rootView.parent as ViewGroup
            viewGroup.removeView(rootView)
        }

        if (useDefaultUiState()) {
            defaultUiController = UiStatusController.get()
            rootView = defaultUiController.bindFragment(rootView)
        }
        return rootView
    }

    /**
     * 使用默认的UiState，即添加在界面的顶层
     * @return
     */
    override fun useDefaultUiState(): Boolean {
        return false
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (useEventBus()) {
            BusProvider.bus.register(this)
        }
        bindEvent()
        initData(savedInstanceState)
    }

    override fun bindUI(rootView: View?) {
        unbinder = KnifeKit.bind(this, rootView!!)
    }

    fun getvDelegate(): VDelegate {
        return vDelegate
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is Activity) {
            this.context = context
        }
    }

    override fun onDetach() {
        super.onDetach()
        //context = null
    }

    override fun useEventBus(): Boolean {
        return false
    }


    override fun onDestroyView() {
        super.onDestroyView()
        if (useEventBus()) {
            BusProvider.bus.unregister(this)
        }
        getvDelegate().destroy()
        //vDelegate = null

        //if (getPresent() != null) {
        //    getPresent()?.detachV()
        //}
        //p = null

    }

    //protected fun getRxPermissions(): RxPermissions? {
    //    rxPermissions.setLogging(CommonConfig.DEV)
    //    return rxPermissions
    //}

    override fun bindEvent() {

    }

    override fun showLoading() {
        if (!activity!!.isFinishing)
            getvDelegate().showLoading("加载中...")
    }

    override fun showLoading(msg: String?) {
        if (activity != null && !activity!!.isFinishing) {
            getvDelegate().showLoading(msg)
        }
    }

    override fun hideLoading() {
        getvDelegate().dismissLoading()
    }

    companion object {

        private const val STATE_SAVE_IS_HIDDEN = "STATE_SAVE_IS_HIDDEN"
    }

    //protected fun getPresent(): P? {
    //    if (p == null) {
    //        p = newP()
    //        if (p != null) {
    //            p?.attachV(this)
    //        }
    //    }
    //    return p
    //}

    override fun newP(): Any? {
        return null
    }
}
